### DISCUSSION TITLE
https://dagger.cloud alpha - feedback welcome

### DISCUSSION THREAD

**Discussion by gerhard** [dagger/dagger#2783] [Ideas]
If you used https://dagger.cloud with `dagger` let us know what worked well and what you wish was better. All & any feedback welcome!

If you are curious to learn more about how [Dagger Cloud](https://dagger.cloud) makes debugging `dagger` issues easier, this doc page explains how to get started: [Debugging with Dagger Cloud](https://docs.dagger.io/1241/dagger-cloud/).

---

@mircubed @marcosnils please edit this as you see fit (maybe starting with this comment). I just needed a link to use in https://github.com/dagger/dagger/blame/main/docs/includes/_dagger-cloud-cta.md#L5 (part of https://github.com/dagger/dagger/pull/2763).

**Comment by gerhard** [comment]
I was using Dagger Cloud yesterday and noticed a few issues:

- run doesn't finish: https://dagger.cloud/runs/e15a94ac-51ec-4aa3-a638-ffc44369978c
- no plan showing if it fails to evaluate: https://dagger.cloud/runs/07e56777-ef54-4dfe-8360-768434140284 (there are a few others, like this one https://dagger.cloud/runs/58ad0726-19b9-4d1f-9893-04f533b61ce2)
- long run commands are not easy to read (line wrap): https://dagger.cloud/runs/4d827557-f493-4d71-a8a5-2fabd867b4a6
- CUE plan coloring after `rm -rf cue.mod/pkg/*` stops working properly: https://dagger.cloud/runs/4d827557-f493-4d71-a8a5-2fabd867b4a6
- showing logs in full-screen, [similar to how Papertrail used to do it](https://www.youtube.com/watch?v=gFFtrzoQEfI), would make it easier to see what is happening here: https://dagger.cloud/runs/c5ab9457-d7c3-4bc9-b2d3-869966b41cef

Happy to share more context. cc @marcosnils @mircubed @slumbering

**Reply by marcosnils** [reply]
> I was using Dagger Cloud yesterday and noticed a few issues:

Looking at your runs, seems like you're using an outdated `devel` dagger version: `https://github.com/dagger/dagger/commit/0c669c58c` 

Mind bumping to the latest release and trying again please? :pray:

**Reply by marcosnils** [reply]
> * long run commands are not easy to read (line wrap): https://dagger.cloud/runs/4d827557-f493-4d71-a8a5-2fabd867b4a6

cc @slumbering UI feedback

> * CUE plan coloring after `rm -rf cue.mod/pkg/*` stops working properly: https://dagger.cloud/runs/4d827557-f493-4d71-a8a5-2fabd867b4a6

yes, this is because there's no current JS `cue` highlighter. We have discussed about this and we're looking for alternatives.


> * showing logs in full-screen, [similar to how Papertrail used to do it](https://www.youtube.com/watch?v=gFFtrzoQEfI), would make it easier to see what is happening here: https://dagger.cloud/runs/c5ab9457-d7c3-4bc9-b2d3-869966b41cef

Yep, @slumbering wanted to do some improvements here as well.

**Comment by TriplEight** [comment]
I've skimmed the doc. There's no info about how to register a (self hosted) runner, how to host a dagger UI (cloud) in case I want to have it self hosted. How does it load-balance multiple jobs between the runners. How to configure the runners. No word about any integration with k8s.
Basically, right now it doesn't seem any different from running images in any CI system.

**Reply by marcosnils** [reply]
> There's no info about how to register a (self hosted) runner, how to host a dagger UI (cloud) in case I want to have it self hosted. How does it load-balance multiple jobs between the runners. How to configure the runners. No word about any integration with k8s.
> Basically, right now it doesn't seem any different from running images in any CI system.

Hey @TriplEight, apologies for the confusion. Dagger cloud is not a CI/CD platform like the ones you're thinking about (gitlab, github actions) where you need to install an agent/runner in your local infrastructure. Our engine is designed to run in any current CI/CD platform, and it posts telemetry information such as caching status, run times, logs, etc to Dagger Cloud so you can keep track of your executions or share them with us in case something's not working as intended. 

Having said that, we're constantly looking for ideas on different features to make Dagger Cloud more useful to our users. If there's something that you'd particularly think it will improve your day to day work, we'd love to hear about it!

**Reply by luolong** [reply]
I was scouring through all the docs and examples I was also similarly confused.

Perhaps a crucial documentation explaining the role of Dagger and its relation to existing CI platforms is missing.

At first (about a year or so ago), when I first learned about Dagger, it looked to me like a nice CI server independent way of describing pipeline steps and a shallow cli on top of that description to run various steps in different existing CI platforms (GitLab Pipelines, GitHub Actions, etc.) I thought, it's neat and I need to take a closer look at it, but at the time my priorities have been elsewhere.

Now, coming back to the project about a year later, I find that CUE language pipeline definition is deprecated in favor of describing your pipelines in more conventional programming languages (I know, CUE API is work in progress), with Go and Python being first apis to be published and more to come.

I am more than a little confused what am I supposed to be doing with Dagger. If this is just a fancy build definitions that you can run locally, how is it better than Gradle or Cargo or Webpack or CMake or any of the ten thousand other similar build tools around there?

If this is more, I just do not see how does this tie in with any of those supported CI platforms and how would I integrate Dagger with a new CI/CD platform of choice?

How am I supposed to be taking advantage of Dagger for a greenfield project?

Let's say, I am going to start a brand new project and I need to set up my CI pipelines. 
So, I define the pipeline using Dagger API. 

What do I need to have to run this pipeline on a CI server somewhere? 
Do I need an existing CI server? 
Can I roll my own based purely on Dagger?
What are my options?

What happens if I have set up a pipeline in one of these CI/CD services and I change the Dagger pipeline significantly -- do I need to redo these mappings in the appropriate CI server definitions as well?

**Reply by marcosnils** [reply]
Hey @luolong! thx for the feedback. Please find some answers to your questions below:

> Perhaps a crucial documentation explaining the role of Dagger and its relation to existing CI platforms is missing.

I agree, given the recent shift on multi-language, we haven't had the time to improve other parts of our docs. A clear guide which clarifies what you're looking for is indeed missing. Tagging @vikram-dagger and @shykes for visibility.


> Now, coming back to the project about a year later, I find that CUE language pipeline definition is deprecated in favor of describing your pipelines in more conventional programming languages (I know, CUE API is work in progress), with Go and Python being first apis to be published and more to come.

Correct. Just to provide a bit of more clarity here, the objective of the project hasn't changed. Dagger's mission has always been to be a programmable CI/CD engine that runs your pipelines in containers. The main difference in the past months is that we've decided to shift towards a multi-language approach instead of only allowing pipelines to be defined in CUE. This decision was purely on user feedback.

Going back to the CUE support, it's not deprecated at all, we're just working on migrating the initial code base to the new architectural changes introduced due to the multi-lang project. CUE will have the same level support as all the other languages.

> I am more than a little confused what am I supposed to be doing with Dagger. If this is just a fancy build definitions that you can run locally, how is it better than Gradle or Cargo or Webpack or CMake or any of the ten thousand other similar build tools around there?

Gradle, Cargo, Webpack and CMake are not CI engines. Imagine Dagger as an evolution of Dockerfiles + your CI pipelines that you can express in your favorite language (programmable) and you can run anywhere. To put it in a concrete example, if you have a Java app that you want to bundle in a container and deploy it to AWS (any service) usually the way to do it is:

- Have a Gradle config to test and build the app
- Have a Dockerfile to containerize the application
- Setup any CI system (Jenkins, Github Actions, etc) to pull code and push it the bundled artifact to be deployed. 

Dagger allows you to bootstrap all these steps, run them locally and in your current CI system. So instead of writing a cumbersome YAML file in your CI that expresses the pipeline described below that **only** runs in your CI, Dagger allows you to express it as code, and run it anywhere.


Hopefully this helped to clarify a bit more what Dagger is and what problem we're trying to solve here. Let us know if it's still not clear.

**Reply by shykes** [reply]
> Hey @luolong! thx for the feedback. Please find some answers to your questions below:
> 
> > Perhaps a crucial documentation explaining the role of Dagger and its relation to existing CI platforms is missing.
> 
> I agree, given the recent shift on multi-language, we haven't had the time to improve other parts of our docs. A clear guide which clarifies what you're looking for is indeed missing. Tagging @vikram-dagger and @shykes for visibility.

Currently we have this:

* [What is Dagger?](https://docs.dagger.io/#what-is-dagger)
* [Who is it for?](https://docs.dagger.io/#who-is-it-for)
* [How does it work?](https://docs.dagger.io/#how-does-it-work)
* [Which SDK should I use?](https://docs.dagger.io/#which-sdk-should-i-use)

@luolong @marcosnils what exactly do you see missing?

**Reply by marcosnils** [reply]
I think the section that we're missing is "Dagger vs other solutions".  I've seen this in some other tools ( [1] [2] [3] [4] )and it helps to understand how it fits (or could fit) in the current architecture

[1]: https://cuelang.org/docs/usecases/datadef/#comparisons
[2]: https://earthly.dev/faq#nix
[3]: https://developer.hashicorp.com/terraform/intro/vs
[4]: https://developer.hashicorp.com/waypoint/docs/intro/vs

**Reply by shykes** [reply]
That makes sense. We had it in the previous docs, and could bring it back. Is that what you had in mind @luolong ?

**Reply by luolong** [reply]
Thanks @shykes and @marcosnils for your replies.

There are few concepts that I feel confusion around. 
I think I understand on the higher level what is the difference between build tool and a CI pipeline, but more practical considerations here are what I am struggling with.

For example, I can see how dagger api and (and CUE pipeline definitions) help me build the pipeline and run it locally.
What remains a mystery to me is how this ties in with an actual CI/CD scenario.

I could see how I can essentially craft a CI/CD pipeline in GitLab that executes various stages of the pipeline in various pipeline staps in GitLab, but it doesn't seem to free me from having to maintain GitLab pipeline manifest in addition to Dagger manifest.

And if I re-work the pipeline in Dagger, I would always have to re-work the GitLab pipeline in parallel. Or is there some magic going on at the Dagger "CI Engine" level, that frees me from having to maintain separate GitLab maping to Dagger pipeline?

**Reply by shykes** [reply]
@luolong yes you do need to maintain a Gitlab manifest in parallel: but that manifest will be much smaller: just enough to run dagger from the right triggers. It also changes much less frequently, since the actual logic is all in your dagger code.

Most of the trouble from using a proprietary CI platform is not the occasional writing of a YAML manifest: it’s the continuous development and testing of complex pipeline logic in YAML. Moving the logic to dagger, while keeping a simple static configuration in Gitlab, is a good separation of concerns.

Down the road you could also bypass the yaml phase entirely, and run dagger from a webhook or git server hook; depending on the capabilities of your git hosting provider.

**Reply by luolong** [reply]
Okay, so, theoretically speaking, CI platforms that allow dynamic generation of pipeline manifests (Like Drone CI or Buildkite for example), could be made to generate entire pipeline manifest on the fly.

Of course, there are pipelines that have different sets of steps based on the trigger type and other attributes, but that is probably whole another can of worms there...

**Reply by shykes** [reply]
> Okay, so, theoretically speaking, CI platforms that allow dynamic generation of pipeline manifests (Like Drone CI or Buildkite for example), could be made to generate entire pipeline manifest on the fly.
> 
> Of course, there are pipelines that have different sets of steps based on the trigger type and other attributes, but that is probably whole another can of worms there...

Sorry, I don't understand what you mean by the above. Does this change your understanding of how Dagger can integrate in your CI runner?

**Reply by luolong** [reply]
> Does this change your understanding of how Dagger can integrate in your CI runner?

No, it dopes not _change_ my understanding. But it does confirm my assumptions.

> Sorry, I don't understand what you mean by the above

So, there exist CI tools ([Buildkite] and [Drone CI]/[Woodpecker CI] that I know of) that allow generating pipelines (or their DAG's) on demand for each pipeline run.

I was simply wondering if Dagger could be used at the pipeline generation phase to more closely match the Dagger specification that is being run. That would allow avoid the necessity of manually keeping the CI platform manifests in sync with Dagger definitions. Just a thought.

[Buildkite]: https://buildkite.com/
[Drone CI]: https://www.drone.io/
[Woodpecker CI]: https://woodpecker-ci.org/

**Comment by sebasjm** [comment]
Can't use `dagger login` from server

```
10:18AM INFO  system | logging into your dagger account
10:18AM INFO  system | opening https://auth.dagger.cloud/authorize?....
10:18AM FATAL system | login failed: exec: "xdg-open,x-www-browser,www-browser": executable file not found in $PATH
```

**Reply by marcosnils** [reply]
thx for the feedback @sebasjm. Expect some updates about this really soon :raised_hands:
